### DISCUSSION TITLE
Quickest way to trigger a berry function from a switch

### DISCUSSION THREAD

**Discussion by markusdd** [arendst/Tasmota#20503] [Q&A]
Situation:
I have a display I wish to turn on when someone comes near it, so I connected a PIR sensor and in general the setup works exactly as intended.

I set GPIO43 to Switch 1, and used `SwitchMode1 15` to decouple the switch from the backlight relay as I only wish to evaluate the 'ON event, turn off time is dependent on other factors and should not be handled by the PIR.

My current solution can be seen below. SwitchMode1 15 causes the switch to emit MQTT messages of the kind `14:38:42.422 MQT: tele/Sunton7-001/SENSOR = {"Time":"2024-01-15T14:38:42","Switch1":"ON"}` which I currently subscribe to. This in turn triggers the pir_sensor function which then emits the screen wakeup event.

This works great and reliably, but it is a tad slow. Between the sensor signal arriving and the screen switching is about 3-4 seconds.
This seems to be because the tele message takes some time to be emitted and then there is some turnaround back from the broker.

Actually, this would not be needed at all because all info is available locally.

In the docs I have found 'tasmota.get_switches', but obviously I do not want to poll that every second. Is there a way I can register a trigger on a purely local event like a switch? (is using switch even the right approach? According to docs it is for PIR but that assumes a relay is couple directly I suppose which is not the case here)

The screen wakeup itself using the touch route is super fast, so if the screen is off and I touch it the wakeup is basically instant, so I assume my wakeup event funtion is fine.

Any advice on how to speed this up is appreciated.

Also, the current subscription would 'swallow' all tele SENSOR messages of the device, which is not what I want, but as the Switch1 is only in the JSON I wonder if one could only subscribe to the Switch1, not the whole SENSOR subtopic.

```
def pir_sensor(topic, idx, payload_s, payload_b)
    if tasmota.get_switches()[0]
        # this comes from an MQTT topic which we generate because of SwitchMode1 15
        # but instead of picking apart JSON we can just check the switch locally
        # if the event was triggered and the switch is on, we wake up the display
        lv.event_send(ui_Screen1, lv.EVENT_CLICKED, 0)
        print("PIR Sensor triggered, sending wakeup event.")
    end
    return true
end
mqtt.subscribe(f"tele/Sunton7-001/SENSOR", pir_sensor)
```

**Comment by sfromis** [comment]
When the PIR is on the same device, you should not at all use `mqtt.subscribe` or even use the SENSOR payload. The better thing to do is something like:
```
def pir_sensor(value, trigger, data)
  lv.event_send(ui_Screen1, lv.EVENT_CLICKED, 0)
  print("PIR Sensor triggered, sending wakeup event.")
end
tasmota.add_rule('Switch1#state`, pir_sensor)
```

**Comment by markusdd** [comment]
ah, via Rules! that is a good idea. I will try that and come back.

**Reply by sfromis** [reply]
What I posted was **not** rules, but a Berry feature with similarites to (not the same as) rules.

**Reply by markusdd** [reply]
interesting. anyway, syntax is identical, but meanwhile I also confirmed Rules are compiled into my build, but unfortunately it does not trigger. Do I have to use a different Switch mode?

**Comment by markusdd** [comment]
mmh. that does not do anything, also does not throw an error. my sneaking suspicion is that in this ESP32-S3 QIO build rules are turned off. I neither see script nor rules in the script menu.
So this probably needs a special build.

**Reply by Jason2866** [reply]
In all esp32 builds rules are enabled. Only when scripting is in the name, guess what's enabled.

**Reply by markusdd** [reply]
I cross-checked and Rules are compiled in indeed.
I am just wondering why I am not seeing the rules menu?

**Comment by markusdd** [comment]
ok, figured it out.

`tasmota.add_rule("Switch1", pir_sensor)` does work. `"Switch1#state"` interestingly does not, I guess that really is Rule Engine syntax.

But anyway, with this the wakeup is basically instant!

Thanks for the input and suggestions!

**Reply by sfromis** [reply]
The syntax I posted very much works. Example:
``` 
tasmota.add_rule("Switch1#state", /value,trigger,data->print(value,trigger,data))
2 Switch1#state {'Switch1': {'State': 2}}
2 Switch1#state {'Switch1': {'State': 2}}
```
This is when using a SwitchMode which does not **only** post to MQTT, but one or the "normal" settings of SwitchMode. Doing it this way is also likely to be *little bit* faster.

**Reply by markusdd** [reply]
can you elaborate what a 'normal' SwitchMode would be?
I used 15 because I needed to detach it from the backlight switch.

In Switchmode 15 Switch1#state does absolutely nothing.
(either this is a bug, or it needs documentation. I do not really see why it would be intended)

Interesting in your printout is also that your value is 2 (a number), while for me I get "ON" or "OFF" (which is also what I want)

**Reply by markusdd** [reply]
What I see from the docs is I could restructure in a sense that I use SwitchMode3, which would trigger on rising edge but ignore the falling:
![grafik](https://github.com/arendst/Tasmota/assets/25175069/1576f9eb-7a80-4083-9e63-dc1bcc731afc)

This works, but would make the berry code a little less clear because you would implicitly need to know that in the background the PIR rising edge is indeed being handled and the falling edge is simply ignored.

Matter of taste I guess, the current version works and is instant in SwitchMode15 which completely detaches it from any background actions.

The remaining question is if SwitchMode15 indeed should not work on Swicth1#state or if this is an unintended bug.

**Reply by sfromis** [reply]
No bug, the docs says this about `SwitchMode 15`: (my emphasis)
> Send **only** MQTT message on switch change. This will stop the switch from controlling power outputs.

The "normal" SwitchMode values are all other than:
> This is when using a SwitchMode which does not **only** post to MQTT

Where the values are available to both "classic" rules and Berry add_rule, as `Switch1#state`

**Reply by markusdd** [reply]
yes but this does not say Switch<x>#state does not work in Rules anymore. At least to me that is not clear.
Because in order to send the messages Tasmota surely still has the state of the switch. Also, if I were to take the 'only' literally, it also should not be emitting the Switch event without state.

If accepted I would propose a small doc PR to make that clear. Because as far as I see 15 (and 16 for inverted) is the only way to detach it fully from local background relay control if that is desired to e.g. do everything in full berry or Rules.
If one were then to copy the example code from the docs or what you gave me above (despite me stating I'm in SwitchMode15) it does not work and it's hard to debug why.

Also poses the question if actually a hand-off switch mode to berry would make sense. Because currently you are forced to pick a SwitchMode that will generate MQTT messages even if you do not need them, spamming the broker with useless stuff.

So Something like 

"
SwitchMode 13 

Detaches the switch from all relays, but does not generate MQTT messages like Mode 15 and 16. This can be used to handle all switch events from either Rules or Berry using "Switch<x>" as the event trigger. 
"

@s-hadinger any opinion from you on this?

**Reply by sfromis** [reply]
Well, the key word is "only", to be read as it does not provide the "normal" action of making a state available. But yeah, that's easier to see when already knowing how the Tasmota logic around Switch input is. 

If you use `SetOption114` (without using the MQTT-only SwitchMode values), you can get the combination of the inputs being detatched from default actions, and still having the `#state` value available, but this option covers all Switch inputs at the same time. If you *combine* `SetOption114` with `SwitchMode 15` the result seems to be that you still get `#state` values (due to `SetOption114`) and the same messages as with `SwitchMode 15` alone. And no, I was not sure about this before making the experiment :smile: 

One weak spot when comparing "classic" rules and Berry `tasmota.add_rule` is that the former will automatically decouple from default actions (without needing either `SwitchMode` 15/16 or `SetOption114`) when all `#state` values are covered, while the latter does not cause such decoupling.

You could get around this without using a MQTT-only `SwitchMode` or `SetOption114` for decoupling, by having a "dummy" classic rule like `on Switch1#state do endon`. Doing it this way cuts down on MQTT messages you may not have any use for, and you get the *slightly* faster processing without depending on the MQTT message JSON.

**Comment by markusdd** [comment]
> If you use `SetOption114` (without using the MQTT-only SwitchMode values), you can get the combination of the inputs being detatched from default actions, and still having the `#state` value available, but this option covers all Switch inputs at the same time. If you _combine_ `SetOption114` with `SwitchMode 15` the result seems to be that you still get `#state` values (due to `SetOption114`) and the same messages as with `SwitchMode 15` alone. And no, I was not sure about this before making the experiment 😄

Ok hahahaha thanks for making this experiment xD

This then frankly really points a little bit to what I was saying above: the behavior is a little inconsistent.
Wouldn't it make sense then to have a SwitchMode 14 (not 114) which does what 114 does but just for a specific switch?
I do not know the history behind these options but it seems a little off to have granual control for each switch in all scenarios but this one scenario can only be applied by a blanket rule.

Frankly: in my very case SetOption114 would also be ok because the backlight is the only relay. But also the current version is fine, I just trigger in Switch1 without state and if the value is "ON" I do whatever I need to do.

Whatever we decide, I think some doc update is definetly in order because it's not really possible to guess this I would argue^^

**Reply by sfromis** [reply]
I'm fine with improvements in docs and/or logic around SwitchMode, but this is not a part of Tasmota which I've used much, and therefore I have limited myself to seeing what happens, instead of expectations about what "should" happen or not, and therefore I've also avoided touching the docs around this. I'm just not sure what was intended, and what was incidental.

**Reply by markusdd** [reply]
understandable. I'm happy to providing a doc PR but which dev could weigh in on what the actual intention is?
